Add nested tag map support

davidlenfesty opened the pull request:
This PR addresses #4 

Currently a draft because of several issues:
1. The testing I've added isn't terribly exhaustive
2.  `set_tag_map` should likely be renamed to `set_global_tag_map` and `set_local_tag_map` could be renamed `set_tag_map`, but I wanted to keep the changes as non-breaking as possible. I'll leave that choice up to you and I can just implement it.
3. There's a small TODO in encoder.py, the implementation as-is may be unsound, needs a little more work/testing to either add proper checks or eliminate the need for them. I'm out of energy for that ATM so I'll come back to it later.

davidlenfesty committed:
fix some whitespace nits

davidlenfesty committed:
add function to set an instance-local tag map

davidlenfesty committed:
Add support for nested tag map configurations

davidlenfesty committed:
add _new_equivalent_tlv function

This is so we can create new TLVs with the same settings applied more
easily

davidlenfesty committed:
Add tests for nested tags

davidlenfesty committed:
Update documentation to include nested tag map functionality

ustropo commented:
About the `set_tag_map`, I think we can create a new function called `set_global_tag_map` and move the implementation to there. Then to just not break anything for now, keep the `set_tag_map` function and call the `set_global_tag_map` inside it. 

Check to see if there's any reference for `set_tag_map` inside the tests and change them to `set_global_tag_map`, as this is the new correct way of doing it.

davidlenfesty committed:
move implementation to set_global_tag_map

davidlenfesty committed:
remove TODO on NestedEncoder parse method

We always pass a TLV to the parse function so it is fine.

davidlenfesty commented:
Better late than never I always say :smiling_face_with_tear:.

I changed the functions to what you wanted (I agree, that's the correct way to do it), and addressed all of my concerns with the PR except for adding tests. There's probably still more testing to be done but I seem to have covered all the code I've added and I can't think of more to add so I'm happy with this.

(side note, [coverage](https://coverage.readthedocs.io/en/6.4.1/) is a pretty neat tool that let me just get a simple coverage report in a couple minutes with no setup)
